nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/ListenSMTP.java

```diff
@@ -0,0 +1,385 @@
+        SSLContext sslContext = null;
+        final SSLContextService sslContextService = context.getProperty(SSL_CONTEXT_SERVICE).asControllerService(SSLContextService.class);
+        if (sslContextService != null) {
+            clientAuth = context.getProperty(CLIENT_AUTH).getValue();
+            sslContext = sslContextService.createSSLContext(SSLContextService.ClientAuth.valueOf(clientAuth));
+        }
+
+        final SSLContext finalSslContext = sslContext;
+
+        SMTPMessageHandlerFactory smtpMessageHandlerFactory = new SMTPMessageHandlerFactory(messages);
+        final SMTPServer server = new SMTPServer(smtpMessageHandlerFactory) {
+
+            @Override
+            public SSLSocket createSSLSocket(Socket socket) throws IOException {
+                InetSocketAddress remoteAddress = (InetSocketAddress) socket.getRemoteSocketAddress();
+
+                SSLSocketFactory socketFactory = finalSslContext.getSocketFactory();
+
+                SSLSocket s = (SSLSocket) (socketFactory.createSocket(socket, remoteAddress.getHostName(), socket.getPort(), true));
+
+                s.setUseClientMode(false);
+
+
+                // For some reason the createSSLContext above is not enough to enforce
+                // client side auth
+                // If client auth is required...
+                if (SSLContextService.ClientAuth.REQUIRED.toString().equals(context.getProperty(CLIENT_AUTH).getValue())) {
+                    s.setNeedClientAuth(true);
+                }
+
+
+                return s;
+            }
+        };
+
+
+        // Set some parameters to our server
+        server.setSoftwareName("Apache NiFi");
+
+
+        // Set the Server options based on properties
+        server.setPort(context.getProperty(SMTP_PORT).asInteger());
+        server.setHostName(context.getProperty(SMTP_HOSTNAME).getValue());
+        server.setMaxMessageSize(context.getProperty(SMTP_MAXIMUM_MSG_SIZE).asDataSize(DataUnit.B).intValue());
+        server.setMaxConnections(context.getProperty(SMTP_MAXIMUM_CONNECTIONS).asInteger());
+        server.setConnectionTimeout(context.getProperty(SMTP_TIMEOUT).asTimePeriod(TimeUnit.MILLISECONDS).intValue());
+
+
+        // Check if TLS should be enabled
+        if (sslContextService != null) {
+            server.setEnableTLS(true);
+        } else {
+            server.setHideTLS(true);
+        }
+
+        // Set TLS to required in case CLIENT_AUTH = required
+        if (SSLContextService.ClientAuth.REQUIRED.toString().equals(context.getProperty(CLIENT_AUTH).getValue())) {
+            server.setRequireTLS(true);
+        }
+
+        this.server = server;
+        server.start();
+
+        getLogger().info("Server started and listening on port " + server.getPort());
+
+        initialized.set(true);
+    }
+
+    @OnStopped
+    public void shutdown() throws Exception {
+        if (server != null) {
+            getLogger().debug("Shutting down server");
+            server.stop();
```

Comment:
Maybe this should be done in an OnUnScheduled function instead of OnStopped (would require adding a Stopping flag for  "initializeSMTPServer" to check). That way the server won't continue to receive messages while the processor is trying to stop and give the OnTrigger the chance to deal with the final messages.